### PR TITLE
[FormRecognizer] Preview 1: argument validation

### LINKED ISSUE #11383: Add argument validation to public methods and constructors on FormRecognizerClient and FormTrainingClient

### PR DESCRIPTION
**Pull request by kinelski**
These changes include the remaining bits of argument validation for the first preview. These include:

- Argument validation for all public methods of the `FormRecognizerClient`.
- `GUID` format validation for the `FormTrainingClient` as well. It's been decided that an `ArgumentException` will be thrown when an invalid format is specified.
- Existing tests that were being ignored are passing now. We don't have skipped tests anymore.
- New tests for the remaining methods.

Fixes #11383.

### CODE REVIEW

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs [hunk 1]
```diff
@@ -199,5 +207,29 @@ public virtual async Task<Response<AccountProperties>> GetAccountPropertiesAsync
...
+
+        /// <summary>
+        /// Used as part of argument validation. Attempts to create a <see cref="Guid"/> from a <c>string</c> and
+        /// throws an <see cref="ArgumentException"/> in case of failure.
+        /// </summary>
+        /// <param name="modelId">The model identifier to be parsed into a <see cref="Guid"/>.</param>
+        /// <param name="paramName">The original parameter name of the <paramref name="modelId"/>. Used to create exceptions in case of failure.</param>
+        /// <returns>The <see cref="Guid"/> instance created from the <paramref name="modelId"/>.</returns>
+        /// <exception cref="ArgumentException">Happens when parsing fails.</exception>
+        private static Guid ValidateModelId(string modelId, string paramName)
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs:219]: Would it make sense to have this in a common place so we don't duplicate code?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs [hunk 2]
```diff
@@ -191,6 +189,126 @@ public void StartRecognizeReceiptsFromUriRespectsTheCancellationToken()
...
+        /// <summary>
+        /// Verifies functionality of the <see cref="FormRecognizerClient.StartRecognizeCustomFormsAsync"/>
+        /// method.
+        /// </summary>
+        [Test]
+        public void StartRecognizeCustomFormsValidatesTheModelIdFormat()
+        {
+            var client = CreateClient();
+            using var stream = new MemoryStream(Array.Empty<byte>());
+            var options = new RecognizeOptions { ContentType = ContentType.Jpeg };
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs:231]: Does content type need to be set for this?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs [hunk 3]
```diff
@@ -191,6 +189,126 @@ public void StartRecognizeReceiptsFromUriRespectsTheCancellationToken()
...
+
+            Assert.ThrowsAsync(expectedType, async () => await client.StartRecognizeCustomFormsFromUriAsync(modelId, uri));
+        }
+
+        /// <summary>
+        /// Verifies functionality of the <see cref="FormRecognizerClient.StartRecognizeCustomFormsFromUriAsync"/>
+        /// method.
+        /// </summary>
+        [Test]
+        public void StartRecognizeCustomFormsFromUriRequiresTheFormFileStream()
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs:277]: Is this name correct?  I'm confused about this requiring the file stream...  should it be requires the URI?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormTrainingClient/FormTrainingClientTests.cs [hunk 4]
```diff
@@ -61,6 +61,7 @@ public void GetCustomModelArgumentValidation()
 
             Assert.ThrowsAsync<ArgumentNullException>(() => client.GetCustomModelAsync(null));
             Assert.ThrowsAsync<ArgumentException>(() => client.GetCustomModelAsync(string.Empty));
+            Assert.ThrowsAsync<ArgumentException>(() => client.GetCustomModelAsync("1975-04-04"));
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormTrainingClient/FormTrainingClientTests.cs:64]: What's special about 1975-04-04?

**Review by annelo-msft — APPROVED** [review]

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs (hunk 2 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs:231]: In fact it's not needed, but I prefer keeping it like this so we can make sure that what's triggering the `ArgumentException` is not the content type validation.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs (hunk 3 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs:277]: Yep. Copy-paste mistakes.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormTrainingClient/FormTrainingClientTests.cs (hunk 4 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormTrainingClient/FormTrainingClientTests.cs:64]: [It's an easter egg!](https://en.wikipedia.org/wiki/Microsoft#1972%E2%80%931985:_The_founding_of_Microsoft)

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs (hunk 2 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/tests/FormRecognizerClient/FormRecognizerClientTests.cs:231]: I had included it in the Uri method as well by accident. I'll remove it.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs (hunk 1 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs]: Discussed offline: the method has been moved to a new `ClientCommon` class.

**Pull request merged**